File: platform/sdk/go/native_program_lifecycle.go
package layerx

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"encoding/json"
)

type NativeProgramDeploy struct {
	ProgramID [32]byte
	GuestABI  uint16
	Policy    byte
	Authority [32]byte
	NewHash   [32]byte
	Interface []byte
	Wasm      []byte
}

type NativeProgramUpgrade struct {
	ProgramID      [32]byte
	GuestABI       uint16
	OldHash        [32]byte
	NewHash        [32]byte
	MigrationHook  []byte
	ClearInterface bool
	Interface      []byte
	Wasm           []byte
}

type NativeProgramWindDown struct {
	ProgramID     [32]byte
	Operation     byte
	Account       [32]byte
	Asset         [32]byte
	Destination   [32]byte
	Seed          []byte
	ExitProgram   [32]byte
	DeadlineBatch uint64
}

func lifecycleInvalid() error { return newSDKError(ErrorInvalidArgument, RetryNever) }

func lifecycleCodeValid(program [32]byte, abi uint16, hash [32]byte, wasm []byte) bool {
	return program != ([32]byte{}) && abi >= 1 && abi <= 3 && len(wasm) >= 8 && len(wasm) <= 1048576 && bytes.Equal(wasm[:8], []byte{0, 97, 115, 109, 1, 0, 0, 0}) && sha256.Sum256(wasm) == hash
}

func EncodeNativeProgramDeploy(value NativeProgramDeploy) ([]byte, error) {
	if !lifecycleCodeValid(value.ProgramID, value.GuestABI, value.NewHash, value.Wasm) || value.Policy > 1 || (value.Policy == 0) != (value.Authority == ([32]byte{})) || len(value.Interface) > 952 || (value.Interface != nil && len(value.Interface) == 0) {
		return nil, lifecycleInvalid()
	}
	fixed := 104
	if value.Interface != nil {
		fixed = 108
	}
	out := make([]byte, fixed)
	copy(out, value.ProgramID[:])
	binary.BigEndian.PutUint16(out[32:], value.GuestABI)
	out[34] = value.Policy
	copy(out[36:], value.Authority[:])
	copy(out[68:], value.NewHash[:])
	binary.BigEndian.PutUint32(out[100:], uint32(len(value.Wasm)))
	if value.Interface != nil {
		binary.BigEndian.PutUint32(out[104:], uint32(len(value.Interface)))
		out = append(out, value.Interface...)
	}
	return append(out, value.Wasm...), nil
}

func DecodeNativeProgramDeploy(payload []byte) (NativeProgramDeploy, error) {
	var value NativeProgramDeploy
	if len(payload) < 104 || payload[35] != 0 {
		return value, lifecycleInvalid()
	}
	copy(value.ProgramID[:], payload[:32])
	value.GuestABI = binary.BigEndian.Uint16(payload[32:])
	value.Policy = payload[34]
	copy(value.Authority[:], payload[36:68])
	copy(value.NewHash[:], payload[68:100])
	wasmLength := uint64(binary.BigEndian.Uint32(payload[100:]))
	offset := 104
	if wasmLength != uint64(len(payload)-104) {
		if len(payload) < 108 {
			return value, lifecycleInvalid()
		}
		length := uint64(binary.BigEndian.Uint32(payload[104:]))
		if length == 0 || length > 952 || 108+length+wasmLength != uint64(len(payload)) {
			return value, lifecycleInvalid()
		}
		value.Interface = append([]byte{}, payload[108:108+int(length)]...)
		offset = 108 + int(length)
	}
	value.Wasm = append([]byte{}, payload[offset:]...)
	encoded, err := EncodeNativeProgramDeploy(value)
	if err != nil || !bytes.Equal(encoded, payload) {
		return NativeProgramDeploy{}, lifecycleInvalid()
	}
	return value, nil
}

func EncodeNativeProgramUpgrade(value NativeProgramUpgrade) ([]byte, error) {
	if !lifecycleCodeValid(value.ProgramID, value.GuestABI, value.NewHash, value.Wasm) || len(value.MigrationHook) > 65535 || len(value.Interface) > 952 || (value.ClearInterface && value.Interface == nil) || (value.Interface != nil && len(value.Interface) == 0 && !value.ClearInterface) {
		return nil, lifecycleInvalid()
	}
	fixed := 106
	if value.Interface != nil {
		fixed = 110
	}
	out := make([]byte, fixed)
	copy(out, value.ProgramID[:])
	binary.BigEndian.PutUint16(out[32:], value.GuestABI)
	if len(value.MigrationHook) != 0 {
		out[34] = 1
	}
	if value.ClearInterface {
		out[34] |= 2
	}
	copy(out[36:], value.OldHash[:])
	copy(out[68:], value.NewHash[:])
	binary.BigEndian.PutUint16(out[100:], uint16(len(value.MigrationHook)))
	binary.BigEndian.PutUint32(out[102:], uint32(len(value.Wasm)))
	if value.Interface != nil {
		binary.BigEndian.PutUint32(out[106:], uint32(len(value.Interface)))
	}
	out = append(out, value.MigrationHook...)
	out = append(out, value.Interface...)
	return append(out, value.Wasm...), nil
}

func DecodeNativeProgramUpgrade(payload []byte) (NativeProgramUpgrade, error) {
	var value NativeProgramUpgrade
	if len(payload) < 106 || payload[35] != 0 || payload[34]&0xfc != 0 {
		return value, lifecycleInvalid()
	}
	copy(value.ProgramID[:], payload[:32])
	value.GuestABI = binary.BigEndian.Uint16(payload[32:])
	copy(value.OldHash[:], payload[36:68])
	copy(value.NewHash[:], payload[68:100])
	value.ClearInterface = payload[34]&2 != 0
	hook := uint64(binary.BigEndian.Uint16(payload[100:]))
	wasm := uint64(binary.BigEndian.Uint32(payload[102:]))
	offset := 106
	if (payload[34]&1 == 0) != (hook == 0) {
		return value, lifecycleInvalid()
	}
	interfaceLength := uint64(0)
	if value.ClearInterface || hook+wasm != uint64(len(payload)-106) {
		if len(payload) < 110 {
			return value, lifecycleInvalid()
		}
		interfaceLength = uint64(binary.BigEndian.Uint32(payload[106:]))
		offset = 110
		if interfaceLength > 952 || (interfaceLength == 0 && !value.ClearInterface) || 110+hook+interfaceLength+wasm != uint64(len(payload)) {
			return value, lifecycleInvalid()
		}
	}
	value.MigrationHook = append([]byte{}, payload[offset:offset+int(hook)]...)
	offset += int(hook)
	if offset-int(hook) == 110 {
		value.Interface = append([]byte{}, payload[offset:offset+int(interfaceLength)]...)
		offset += int(interfaceLength)
	}
	value.Wasm = append([]byte{}, payload[offset:]...)
	encoded, err := EncodeNativeProgramUpgrade(value)
	if err != nil || !bytes.Equal(encoded, payload) {
		return NativeProgramUpgrade{}, lifecycleInvalid()
	}
	return value, nil
}

func EncodeNativeProgramWindDown(value NativeProgramWindDown) ([]byte, error) {
	if value.ProgramID == ([32]byte{}) {
		return nil, lifecycleInvalid()
	}
	out := append([]byte{}, value.ProgramID[:]...)
	out = append(out, value.Operation)
	switch value.Operation {
	case 1:
		if len(value.Seed) > 128 || value.ExitProgram != ([32]byte{}) || value.DeadlineBatch != 0 {
			return nil, lifecycleInvalid()
		}
		out = append(out, value.Account[:]...)
		out = append(out, value.Asset[:]...)
		out = append(out, value.Destination[:]...)
		out = binary.BigEndian.AppendUint16(out, uint16(len(value.Seed)))
		out = append(out, value.Seed...)
	case 2:
		if value.Account != ([32]byte{}) || value.Asset != ([32]byte{}) || value.Destination != ([32]byte{}) || len(value.Seed) != 0 {
			return nil, lifecycleInvalid()
		}
		out = append(out, value.ExitProgram[:]...)
		out = binary.BigEndian.AppendUint64(out, value.DeadlineBatch)
	case 3, 4:
		if value.Asset != ([32]byte{}) || value.Destination != ([32]byte{}) || value.ExitProgram != ([32]byte{}) || value.DeadlineBatch != 0 || len(value.Seed) != 0 || (value.Operation == 3 && value.Account != ([32]byte{})) {
			return nil, lifecycleInvalid()
		}
		if value.Operation == 4 {
			out = append(out, value.Account[:]...)
		}
	default:
		return nil, lifecycleInvalid()
	}
	return out, nil
}

func DecodeNativeProgramWindDown(payload []byte) (NativeProgramWindDown, error) {
	var value NativeProgramWindDown
	if len(payload) < 33 {
		return value, lifecycleInvalid()
	}
	copy(value.ProgramID[:], payload[:32])
	value.Operation = payload[32]
	switch value.Operation {
	case 1:
		if len(payload) < 131 || int(binary.BigEndian.Uint16(payload[129:]))+131 != len(payload) {
			return value, lifecycleInvalid()
		}
		copy(value.Account[:], payload[33:65])
		copy(value.Asset[:], payload[65:97])
		copy(value.Destination[:], payload[97:129])
		value.Seed = append([]byte{}, payload[131:]...)
	case 2:
		if len(payload) != 73 {
			return value, lifecycleInvalid()
		}
		copy(value.ExitProgram[:], payload[33:65])
		value.DeadlineBatch = binary.BigEndian.Uint64(payload[65:])
	case 3:
		if len(payload) != 33 {
			return value, lifecycleInvalid()
		}
	case 4:
		if len(payload) != 65 {
			return value, lifecycleInvalid()
		}
		copy(value.Account[:], payload[33:])
	default:
		return value, lifecycleInvalid()
	}
	encoded, err := EncodeNativeProgramWindDown(value)
	if err != nil || !bytes.Equal(encoded, payload) {
		return NativeProgramWindDown{}, lifecycleInvalid()
	}
	return value, nil
}

type NativeProgramLifecycleRequest struct {
	ordinal        uint16
	payload        []byte
	signedActivity []byte
	binding        programCallBinding
}

func NewNativeProgramLifecycleRequest(ordinal uint16, payload, signedActivity []byte) (NativeProgramLifecycleRequest, error) {
	var err error
	switch ordinal {
	case 1:
		_, err = DecodeNativeProgramDeploy(payload)
	case 2:
		_, err = DecodeNativeProgramUpgrade(payload)
	case 7:
		_, err = DecodeNativeProgramWindDown(payload)
	default:
		err = lifecycleInvalid()
	}
	if err != nil {
		return NativeProgramLifecycleRequest{}, err
	}
	binding, err := bindNativeProgramActivity(ordinal, payload, signedActivity)
	if err != nil {
		return NativeProgramLifecycleRequest{}, err
	}
	return NativeProgramLifecycleRequest{ordinal, append([]byte{}, payload...), append([]byte{}, signedActivity...), binding}, nil
}

func bindNativeProgramActivity(ordinal uint16, expected, signed []byte) (programCallBinding, error) {
	invalid := func() (programCallBinding, error) { return programCallBinding{}, lifecycleInvalid() }
	if len(signed) == 0 || len(signed) > 1048576 {
		return invalid()
	}
	decoder := wireDecoder{value: signed}
	if decoder.u16() != 3 || decoder.u16() != 0x1001 || decoder.u8() != 12 || decoder.u8() != 1 || decoder.u16() != 3 || decoder.u8() != 2 {
		return invalid()
	}
	_ = decoder.u32()
	if decoder.u8() != 3 || decoder.u32() != 0x00090000|uint32(ordinal) || decoder.u8() != 4 {
		return invalid()
	}
	_ = decoder.bounded(255)
	if decoder.u8() != 5 {
		return invalid()
	}
	_ = decoder.bounded(524288)
	if decoder.u8() != 6 {
		return invalid()
	}
	_ = decoder.u64()
	if decoder.u8() != 7 {
		return invalid()
	}
	before := decoder.u64()
	after := decoder.u64()
	if after < before || decoder.u8() != 8 {
		return invalid()
	}
	key := decoder.array32()
	if decoder.u8() != 9 {
		return invalid()
	}
	_ = decoder.u128()
	if decoder.u8() != 10 {
		return invalid()
	}
	hash := decoder.array32()
	if decoder.u8() != 11 {
		return invalid()
	}
	payload := decoder.bounded(524288)
	if decoder.u8() != 12 {
		return invalid()
	}
	_ = decoder.bounded(128)
	if decoder.failed || decoder.offset != len(signed) || !bytes.Equal(payload, expected) || domainDigest([]byte("LXP/v1/payload-hash\x00"), payload) != hash {
		return invalid()
	}
	return programCallBinding{ActivityID: domainDigest([]byte("LXP/v1/activity-id\x00"), signed), IdempotencyKey: key, NotBefore: before, NotAfter: after}, nil
}

func (programs *Programs) Deploy(ctx context.Context, value NativeProgramDeploy, signed []byte, key IdempotencyKey) (VerifiedReceipt, error) {
	payload, err := EncodeNativeProgramDeploy(value)
	if err != nil {
		return VerifiedReceipt{}, err
	}
	return programs.lifecycle(ctx, 1, payload, signed, key)
}

func (programs *Programs) Upgrade(ctx context.Context, value NativeProgramUpgrade, signed []byte, key IdempotencyKey) (VerifiedReceipt, error) {
	payload, err := EncodeNativeProgramUpgrade(value)
	if err != nil {
		return VerifiedReceipt{}, err
	}
	return programs.lifecycle(ctx, 2, payload, signed, key)
}

func (programs *Programs) WindDown(ctx context.Context, value NativeProgramWindDown, signed []byte, key IdempotencyKey) (VerifiedReceipt, error) {
	payload, err := EncodeNativeProgramWindDown(value)
	if err != nil {
		return VerifiedReceipt{}, err
	}
	return programs.lifecycle(ctx, 7, payload, signed, key)
}

func (programs *Programs) lifecycle(ctx context.Context, ordinal uint16, payload, signed []byte, key IdempotencyKey) (VerifiedReceipt, error) {
	request, err := NewNativeProgramLifecycleRequest(ordinal, payload, signed)
	if err != nil {
		return VerifiedReceipt{}, err
	}
	if programs.protocolVersion != 3 || !canonicalProgramKey(key) || key.String() != hex.EncodeToString(request.binding.IdempotencyKey[:]) {
		return VerifiedReceipt{}, lifecycleInvalid()
	}
	operation := map[uint16]string{1: "program.deploy", 2: "program.upgrade", 7: "program.wind-down"}[ordinal]
	var raw json.RawMessage
	err = programs.client.Agent(ctx, AgentOperation(operation), map[string]string{"signed_activity": hex.EncodeToString(request.signedActivity), "payload": hex.EncodeToString(request.payload)}, &raw, CallOptions{IdempotencyKey: key})
	if err != nil {
		return VerifiedReceipt{}, err
	}
	return decodeLifecycleSubmission(raw, request.binding.ActivityID, programs.trustedSequencerPublicKey)
}

func decodeLifecycleSubmission(raw json.RawMessage, activity, sequencer [32]byte) (VerifiedReceipt, error) {
	invalid := func() (VerifiedReceipt, error) {
		return VerifiedReceipt{}, newSDKError(ErrorUnknownOutcome, RetryUnknownOutcome)
	}
	var fields map[string]json.RawMessage
	if decodeStrict(raw, &fields) != nil || !exactFields(fields, "state", "activity_id", "receipt", "terminal_payload", "call_graph") {
		return invalid()
	}
	values := map[string]string{}
	for name, field := range fields {
		var value string
		if json.Unmarshal(field, &value) != nil {
			return invalid()
		}
		values[name] = value
	}
	if values["activity_id"] != hex.EncodeToString(activity[:]) || values["terminal_payload"] != "" || values["call_graph"] != "" || (values["state"] != "executed" && values["state"] != "refused") || !canonicalLowerHex(values["receipt"], len(values["receipt"])/2) {
		return invalid()
	}
	canonical, err := hex.DecodeString(values["receipt"])
	if err != nil {
		return invalid()
	}
	verified, err := VerifyProgramLifecycleReceipt(canonical, activity, sequencer)
	if err != nil || (values["state"] == "executed") != (verified.Receipt.ResultCode == 0) {
		return invalid()
	}
	return verified, nil
}

func (programs *Programs) LifecycleReceipt(ctx context.Context, request NativeProgramLifecycleRequest) (VerifiedReceipt, error) {
	bound, err := NewNativeProgramLifecycleRequest(request.ordinal, request.payload, request.signedActivity)
	if err != nil || programs.protocolVersion != 3 {
		return VerifiedReceipt{}, lifecycleInvalid()
	}
	key := hex.EncodeToString(bound.binding.IdempotencyKey[:])
	activity := hex.EncodeToString(bound.binding.ActivityID[:])
	raw, err := programs.raw(ctx, "program.receipt", false, map[string]string{"idempotency_key": key, "expected_activity_id": activity, "requested_verification_level": "sequencer-signed"}, CallOptions{PathParameters: map[string]string{"idempotency_key": key}})
	if err != nil {
		return VerifiedReceipt{}, err
	}
	var fields map[string]json.RawMessage
	var actual, receipt string
	if decodeStrict(raw, &fields) != nil || !exactFields(fields, "activity_id", "receipt") || json.Unmarshal(fields["activity_id"], &actual) != nil || actual != activity || json.Unmarshal(fields["receipt"], &receipt) != nil || len(receipt) > 2097152 || !canonicalLowerHex(receipt, len(receipt)/2) {
		return VerifiedReceipt{}, newSDKError(ErrorVerificationFailure, RetryNever)
	}
	canonical, err := hex.DecodeString(receipt)
	if err != nil {
		return VerifiedReceipt{}, err
	}
	return VerifyProgramLifecycleReceipt(canonical, bound.binding.ActivityID, programs.trustedSequencerPublicKey)
}

File: platform/sdk/dotnet/NativeProgramLifecycle.cs
using System.Buffers.Binary;
using System.Security.Cryptography;
using System.Text;

namespace LayerX.Sdk;

public interface INativeProgramLifecycle
{
    ushort Ordinal { get; }
    byte[] Encode();
}

public sealed class NativeProgramDeploy : INativeProgramLifecycle
{
    private readonly byte[] _payload;
    public ushort Ordinal => 1;
    public NativeProgramDeploy(byte[] programId, ushort guestAbi, byte policy, byte[] authority,
        byte[] newHash, byte[]? programInterface, byte[] wasm)
    {
        LifecycleWire.Code(programId, guestAbi, newHash, wasm); LifecycleWire.Bytes32(authority);
        LifecycleWire.Require(policy <= 1 && (policy == 0) == authority.All(value => value == 0) &&
            (programInterface is null || programInterface.Length is > 0 and <= 952));
        using var output = new MemoryStream();
        output.Write(programId); LifecycleWire.Put(output, guestAbi, 2); output.WriteByte(policy); output.WriteByte(0);
        output.Write(authority); output.Write(newHash); LifecycleWire.Put(output, (ulong)wasm.Length, 4);
        if (programInterface is not null) { LifecycleWire.Put(output, (ulong)programInterface.Length, 4); output.Write(programInterface); }
        output.Write(wasm); _payload = output.ToArray();
    }
    public byte[] Encode() => _payload.ToArray();
    public static NativeProgramDeploy Decode(byte[] payload)
    {
        LifecycleWire.Require(payload.Length >= 104 && payload[35] == 0);
        var wasm = LifecycleWire.Get(payload, 100, 4); var offset = 104; byte[]? programInterface = null;
        if (wasm != (ulong)(payload.Length - 104))
        {
            LifecycleWire.Require(payload.Length >= 108); var length = LifecycleWire.Get(payload, 104, 4);
            LifecycleWire.Require(length is > 0 and <= 952 && 108 + length + wasm == (ulong)payload.Length);
            offset = 108 + (int)length; programInterface = payload[108..offset];
        }
        var value = new NativeProgramDeploy(payload[..32], (ushort)LifecycleWire.Get(payload, 32, 2), payload[34],
            payload[36..68], payload[68..100], programInterface, payload[offset..]);
        LifecycleWire.Require(value.Encode().SequenceEqual(payload)); return value;
    }
}

public sealed class NativeProgramUpgrade : INativeProgramLifecycle
{
    private readonly byte[] _payload;
    public ushort Ordinal => 2;
    public NativeProgramUpgrade(byte[] programId, ushort guestAbi, byte[] oldHash, byte[] newHash,
        byte[] migrationHook, bool clearInterface, byte[]? programInterface, byte[] wasm)
    {
        LifecycleWire.Code(programId, guestAbi, newHash, wasm); LifecycleWire.Bytes32(oldHash);
        LifecycleWire.Require(migrationHook.Length <= ushort.MaxValue && (!clearInterface || programInterface is not null) &&
            (programInterface is null || programInterface.Length <= 952 && (programInterface.Length > 0 || clearInterface)));
        using var output = new MemoryStream();
        output.Write(programId); LifecycleWire.Put(output, guestAbi, 2);
        output.WriteByte((byte)((migrationHook.Length > 0 ? 1 : 0) | (clearInterface ? 2 : 0))); output.WriteByte(0);
        output.Write(oldHash); output.Write(newHash); LifecycleWire.Put(output, (ulong)migrationHook.Length, 2); LifecycleWire.Put(output, (ulong)wasm.Length, 4);
        if (programInterface is not null) LifecycleWire.Put(output, (ulong)programInterface.Length, 4);
        output.Write(migrationHook); if (programInterface is not null) output.Write(programInterface);
        output.Write(wasm); _payload = output.ToArray();
    }
    public byte[] Encode() => _payload.ToArray();
    public static NativeProgramUpgrade Decode(byte[] payload)
    {
        LifecycleWire.Require(payload.Length >= 106 && payload[35] == 0 && (payload[34] & 0xfc) == 0);
        var hook = (int)LifecycleWire.Get(payload, 100, 2); var wasm = LifecycleWire.Get(payload, 102, 4);
        var clear = (payload[34] & 2) != 0; var offset = 106; var length = 0; byte[]? programInterface = null;
        LifecycleWire.Require(((payload[34] & 1) == 0) == (hook == 0));
        if (clear || (ulong)hook + wasm != (ulong)(payload.Length - 106))
        {
            LifecycleWire.Require(payload.Length >= 110); var size = LifecycleWire.Get(payload, 106, 4);
            LifecycleWire.Require(size <= 952 && (size > 0 || clear) && 110 + (ulong)hook + size + wasm == (ulong)payload.Length);
            offset = 110; length = (int)size; programInterface = payload[(offset + hook)..(offset + hook + length)];
        }
        var value = new NativeProgramUpgrade(payload[..32], (ushort)LifecycleWire.Get(payload, 32, 2), payload[36..68], payload[68..100],
            payload[offset..(offset + hook)], clear, programInterface, payload[(offset + hook + length)..]);
        LifecycleWire.Require(value.Encode().SequenceEqual(payload)); return value;
    }
}

public sealed class NativeProgramWindDown : INativeProgramLifecycle
{
    private readonly byte[] _payload;
    private NativeProgramWindDown(byte[] payload) { _payload = payload.ToArray(); }
    public ushort Ordinal => 7;
    public byte Operation => _payload[32];
    public byte[] Encode() => _payload.ToArray();
    public static NativeProgramWindDown Route(byte[] program, byte[] account, byte[] asset, byte[] destination, byte[] seed)
    {
        LifecycleWire.Bytes32(program); LifecycleWire.Bytes32(account); LifecycleWire.Bytes32(asset); LifecycleWire.Bytes32(destination);
        LifecycleWire.Require(seed.Length <= 128); using var output = new MemoryStream();
        output.Write(program); output.WriteByte(1); output.Write(account); output.Write(asset); output.Write(destination);
        LifecycleWire.Put(output, (ulong)seed.Length, 2); output.Write(seed); return Decode(output.ToArray());
    }
    public static NativeProgramWindDown Deprecate(byte[] program, byte[] exitProgram, ulong deadlineBatch)
    {
        LifecycleWire.Bytes32(program); LifecycleWire.Bytes32(exitProgram); using var output = new MemoryStream();
        output.Write(program); output.WriteByte(2); output.Write(exitProgram); LifecycleWire.Put(output, deadlineBatch, 8); return Decode(output.ToArray());
    }
    public static NativeProgramWindDown Tombstone(byte[] program)
    {
        LifecycleWire.Bytes32(program); return Decode([.. program, 3]);
    }
    public static NativeProgramWindDown Exit(byte[] program, byte[] account)
    {
        LifecycleWire.Bytes32(program); LifecycleWire.Bytes32(account); return Decode([.. program, 4, .. account]);
    }
    public static NativeProgramWindDown Decode(byte[] payload)
    {
        LifecycleWire.Require(payload.Length >= 33 && payload[..32].Any(value => value != 0));
        switch (payload[32])
        {
            case 1:
                LifecycleWire.Require(payload.Length >= 131); var seed = LifecycleWire.Get(payload, 129, 2);
                LifecycleWire.Require(seed <= 128 && (ulong)payload.Length == 131 + seed); break;
            case 2: LifecycleWire.Require(payload.Length == 73); break;
            case 3: LifecycleWire.Require(payload.Length == 33); break;
            case 4: LifecycleWire.Require(payload.Length == 65); break;
            default: throw new ArgumentException("Wind-down operation");
        }
        return new NativeProgramWindDown(payload);
    }
}

public sealed class NativeProgramLifecycleRequest
{
    private readonly byte[] _payload;
    private readonly byte[] _signed;
    public ushort Ordinal { get; }
    public byte[] Payload => _payload.ToArray();
    public byte[] SignedActivity => _signed.ToArray();
    public byte[] ActivityId => LifecycleWire.Hash(Encoding.UTF8.GetBytes("LXP/v1/activity-id\0"), _signed);
    public string IdempotencyKey { get; }
    public NativeProgramLifecycleRequest(INativeProgramLifecycle operation, byte[] signedActivity)
    {
        Ordinal = operation.Ordinal; _payload = LifecycleWire.Decode(Ordinal, operation.Encode()).Encode(); _signed = signedActivity.ToArray();
        IdempotencyKey = Convert.ToHexString(LifecycleWire.Bind(Ordinal, _payload, _signed)).ToLowerInvariant();
    }
}

internal static class LifecycleWire
{
    internal static void Require(bool valid) { if (!valid) throw new ArgumentException("Non-canonical Programs lifecycle"); }
    internal static void Bytes32(byte[] value) { Require(value.Length == 32); }
    internal static byte[] Hash(params byte[][] parts)
    {
        using var hash = IncrementalHash.CreateHash(HashAlgorithmName.SHA256);
        foreach (var part in parts) hash.AppendData(part); return hash.GetHashAndReset();
    }
    internal static void Code(byte[] program, ushort abi, byte[] hash, byte[] wasm)
    {
        Bytes32(program); Bytes32(hash);
        Require(program.Any(value => value != 0) && abi is >= 1 and <= 3 && wasm.Length is >= 8 and <= 1048576 &&
            wasm.AsSpan(0, 8).SequenceEqual(new byte[] { 0, 97, 115, 109, 1, 0, 0, 0 }) && CryptographicOperations.FixedTimeEquals(hash, SHA256.HashData(wasm)));
    }
    internal static ulong Get(byte[] bytes, int offset, int length)
    {
        Require(offset >= 0 && length >= 0 && offset <= bytes.Length - length); ulong result = 0;
        foreach (var value in bytes.AsSpan(offset, length)) result = (result << 8) | value; return result;
    }
    internal static void Put(Stream output, ulong value, int length)
    {
        Span<byte> bytes = stackalloc byte[8]; BinaryPrimitives.WriteUInt64BigEndian(bytes, value); output.Write(bytes[(8 - length)..]);
    }
    internal static INativeProgramLifecycle Decode(ushort ordinal, byte[] payload) => ordinal switch
    {
        1 => NativeProgramDeploy.Decode(payload),
        2 => NativeProgramUpgrade.Decode(payload),
        7 => NativeProgramWindDown.Decode(payload),
        _ => throw new ArgumentException("Programs lifecycle ordinal"),
    };
    internal static byte[] Bind(ushort ordinal, byte[] expected, byte[] signed)
    {
        Require(signed.Length is > 0 and <= 1048576); var offset = 0;
        ulong Read(int length) { var value = Get(signed, offset, length); offset += length; return value; }
        byte[] Bounded(int maximum)
        {
            var length = Read(4); Require(length <= (ulong)maximum && length <= (ulong)(signed.Length - offset));
            var value = signed[offset..(offset + (int)length)]; offset += (int)length; return value;
        }
        Require(Read(2) == 3 && Read(2) == 0x1001 && Read(1) == 12 && Read(1) == 1 && Read(2) == 3 && Read(1) == 2);
        Read(4); Require(Read(1) == 3 && Read(4) == (0x00090000UL | ordinal) && Read(1) == 4);
        Bounded(255); Require(Read(1) == 5); Bounded(524288); Require(Read(1) == 6); Read(8); Require(Read(1) == 7);
        var before = Read(8); var after = Read(8); Require(after >= before && Read(1) == 8);
        var key = Bounded(32); Require(key.Length == 32 && Read(1) == 9); Read(8); Read(8); Require(Read(1) == 10);
        var hash = Bounded(32); Require(hash.Length == 32 && Read(1) == 11); var payload = Bounded(524288);
        Require(Read(1) == 12); Bounded(128);
        Require(offset == signed.Length && payload.SequenceEqual(expected) && hash.SequenceEqual(Hash(Encoding.UTF8.GetBytes("LXP/v1/payload-hash\0"), payload)));
        return key;
    }
}
